lib/Dialect/FIRRTL/Transforms/WireDFT.cpp

```diff
@@ -466,30 +487,57 @@ void WireDFTPass::runOnOperation() {
       return arg;
     };
 
-    // Wire the signal to each clock gate using the helper above.
-    for (auto *instance : targets) {
-      auto *parent = instance->getParent();
-      auto module = cast<FModuleOp>(*parent->getModule());
-      auto builder = ImplicitLocOpBuilder::atBlockEnd(module->getLoc(),
-                                                      module.getBodyBlock());
-      emitConnect(
-          builder,
-          cast<InstanceOp>(*instance->getInstance()).getResult(targetPortNo),
-          getSignal(parent));
+    // Wire the signal to each provided destination.
+    // Insert connect at end, to ensure this "wins" re:last-connect.
+    // This may not be legal if destination is not top-level
+    // but this sort of assumption is made throughout and hasn't
+    // been a problem yet :(.
+    for (Value dest : dests) {
+      auto *op = dest.getDefiningOp();
+      assert(op && "block arguments not supported presently");
+      auto parent = op->getParentOfType<FModuleOp>();
+      assert(parent);
+      auto builder =
+          ImplicitLocOpBuilder::atBlockEnd(op->getLoc(), parent.getBodyBlock());
+      emitConnect(builder, dest, getSignal(instanceGraph.lookup(parent)));
     }
+
     return success();
   };
 
+  // Gather all test-enable targets: test-enable result of instance op's,
+  // as well as wires used as test-enable operand of the intrinsics.
+  // Once instance variant is deprecated, just feed list of intrinsics to the
+  // helper instead.
+  auto testEnableDests = llvm::map_to_vector(clockGates, [&](auto *rec) {
+    return cast<InstanceOp>(*rec->getInstance()).getResult(testEnPortNo);
+  });
+  llvm::append_range(
+      testEnableDests,
+      llvm::map_range(cgIntrinsics, [&](ClockGateIntrinsicOp cgOp) {
+        // Insert wire to run test-enable signal to,
+        // and change the CG intrinsic to use it as such.
+        auto builder = ImplicitLocOpBuilder(cgOp.getLoc(), cgOp);
+        auto wire = builder.create<WireOp>(enableSignal.getType());
+        cgOp.getTestEnableMutable().assign(wire.getResult());
+        return wire.getResult();
+      }));
```

Comment:
Clever! I was wondering how you're going to deal with the intrinsic having an actual operand instead of a wire-y thing. This is a nice low-friction way to do it!

Reply:
Yeah, it will add unnecessary wires in common case where the test-enable comes from above, but needed for some cases where it comes from below or in same module (without dominance checking).  The wiring logic here isn't as robust as WiringProblem's and that has some limitations too IIRC in some corner cases.  Anyway, once we don't have to handle both forms it'd be good to rework this to handle this better but low-friction indeed :+1: .